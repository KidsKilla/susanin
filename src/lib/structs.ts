import { assocPath, dissocPath } from 'ramda';
import { v4 as uuid } from 'uuid';

export enum TVertexType {
    Normal,
    Boulder,
    Gravel,
}
export type TGraphLinksTreeNode = { [id: string]: number };
export type TGraphLinksTree = { [id: string]: TGraphLinksTreeNode };
export type TVertexId = string;
export type TVertex = {
    id: TVertexId,
    type: TVertexType,
};
export type TWormholeDict = { [id: string]: TVertexId };
type TVertexDict = { [id: string]: TVertex };
export type TGraph = {
    wormholes: TWormholeDict,
    vertices: TVertexDict,
    board: TBoard,
};
export type TBoard = TVertexId[][];

export const createVertex = (type: TVertexType = TVertexType.Normal, id: string = uuid()): TVertex => ({ id, type });

const createArray = (size: number): string[] => [...new Array(size).join(' ').split(' ')];
const TYPE_MAP = {
    '.': TVertexType.Normal,
    'G': TVertexType.Gravel,
    '#': TVertexType.Boulder,
};
type TRevDict = { [id: string]: string };
const TYPE_MAP_REV: TRevDict = {
    [TVertexType.Normal]: '.',
    [TVertexType.Gravel]: 'G',
    [TVertexType.Boulder]: '#',
};

export const getIdByCoords = (board: TBoard, x: number, y: number): TVertexId | undefined =>
    board[y] ? board[y][x] : undefined;

export const createGraphFromStringMap = (map: string, whList: number[][][]): TGraph => {
    const vertices: TVertexDict = {};
    const board: TBoard = map.trim().split('\n')
        .map(
            (line: string, y: number) => line
                .trim()
                .split('')
                .map((type: string, x: number) => {
                    const vType = type in TYPE_MAP ? TYPE_MAP[type] : TVertexType.Boulder;
                    const vertex = createVertex(vType, JSON.stringify({ x, y, _: type }));
                    vertices[vertex.id] = vertex;
                    return vertex.id;
                })
        );

    const wormholes = whList.reduce(
        (memo, [whFrom, whTo]) => {
            const idFrom = getIdByCoords(board, whFrom[0], whFrom[1]);
            const idTo = getIdByCoords(board, whTo[0], whTo[1]);
            if (idFrom && idTo) {
                memo[idFrom] = idTo;
            }
            return memo;
        },
        {}
    );

    return {
        wormholes,
        board,
        vertices,
    };
};

export const getStringMapFromGraph = (graph: TGraph): string => {
    const lines: string[] = graph.board
        .map(xLine => xLine
            .map(vertexId => {
                const vertex = graph.vertices[vertexId];
                return TYPE_MAP_REV[vertex.type];
            })
            .join('')
        );
    return ['', ...lines, ''].join('\n');
};

export const createEmptyGraph = (xSize: number, ySize: number): TGraph => {
    const vertices = {};
    const board = createArray(ySize)
        .map((_1, y: number) => {
            let line = createArray(xSize);
            return line.map((_2, x: number) => {
                const vertexId = JSON.stringify({ x, y });
                const vertex = createVertex(TVertexType.Normal, vertexId);
                vertices[vertex.id] = vertex;
                return vertex.id;
            });
        });

    return {
        wormholes: {},
        vertices,
        board,
    };
};

const updateArrayLength = (size: number, array: Array<string | string[]>): Array<string | string[]> => {
        if (size > array.length) {
            return array.concat(createArray(size - array.length));
        } else if (size < array.length) {
            const newArray = [...array];
            newArray.length = size;
            return newArray;
        }
        return array;
    };

const traverseBoard = (callback: Function, board: TBoard): void => {
    board.forEach((line: TVertexId[], y: number) => {
        line.forEach((vertexId: TVertexId, x: number) => {
            callback(vertexId, x, y);
        });
    });
};

export const updateBoardSize = (xSize: number, ySize: number, graph: TGraph): TGraph => {
    const vertices = { ...graph.vertices };

    // update board
    const board = updateArrayLength(ySize, graph.board)
        .map((line: string[], y: number) =>
            updateArrayLength(xSize, line || createArray(xSize))
                .map((oldVertexId: TVertexId, x: number) => {
                    let id = JSON.stringify({ x, y });
                    const vertex = oldVertexId
                        ? vertices[oldVertexId]
                        : createVertex(TVertexType.Normal, id);
                    vertices[vertex.id] = vertex;
                    return vertex.id;
                })
        );

    // remove deleted vertices
    traverseBoard(
        (vertexId: TVertexId, x: number, y: number) => {
            if (getIdByCoords(board, x, y) === undefined) {
                delete vertices[vertexId];
            }
        },
        graph.board
    );

    // remove invalid wormholes
    const wormholes = { ...graph.wormholes };
    Object.keys(graph.wormholes).forEach((whInId: TVertexId) => {
        const whOutId = graph.wormholes[whInId];
        if (!vertices[whInId] || !vertices[whOutId]) {
            delete wormholes[whInId];
        }
    });

    const newGraph: TGraph = {
        board,
        wormholes,
        vertices,
    };
    return newGraph;
};

const vtxPath = (vertexId: TVertexId): string[] => ['vertices', vertexId, 'type'];
const whPath = (vertexId: TVertexId) => ['wormholes', vertexId];

export const updateGraphVertexById = (type: TVertexType, id: TVertexId, graph: TGraph): TGraph =>
    assocPath(vtxPath(id), type, graph);

export const updateGraphVertexByCoords = (type: TVertexType, x: number, y: number, graph: TGraph): TGraph => {
    const id = getIdByCoords(graph.board, x, y);
    return id ? updateGraphVertexById(type, id, graph) : graph;
};

export const addGraphWormhole = (from: TVertexId, to: TVertexId, graph: TGraph): TGraph =>
    assocPath(whPath(from), to, graph);

export const removeGraphWormhole = (vertexId: TVertexId, graph: TGraph): TGraph => {
    if (!graph.wormholes[vertexId]) {
        Object.keys(graph.wormholes).some(whInId => {
            let whOutId = graph.wormholes[whInId];
            const isFound = whOutId === vertexId;
            if (isFound) {
                vertexId = whInId;
            }
            return isFound;
        });
    }
    if (!graph.wormholes[vertexId]) {
        throw new Error(`Vertex is not in graph: ${vertexId}`);
    }
    return dissocPath(whPath(vertexId), graph);
};
