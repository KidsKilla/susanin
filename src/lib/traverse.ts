import { TGraph, TGraphLinksTree, TVertex, TVertexId, TVertexType } from './structs';

type TCosts = object;

const getLowestCostNode = (costs: TCosts, processed: TVertexId[]): TVertexId => Object.keys(costs)
    .reduce(
        (cheapVertexId: TVertexId, vertexId: TVertexId) => {
            const isCheaper = cheapVertexId === '' || costs[vertexId] < costs[cheapVertexId];
            const isNew = processed.indexOf(vertexId) === -1;
            if (isCheaper && isNew) {
                cheapVertexId = vertexId;
            }
            return cheapVertexId;
        },
        ''
    );

type TParents = { [id: string]: string };
const getOptimalPath = (parents: TParents, finishId: string) => {
    const optimalPath = [];
    let parent = finishId;
    do {
        optimalPath.push(parent);
        parent = parents[parent];
    } while (parent);
    optimalPath.reverse();
    return optimalPath;
};

// function that returns the minimum cost and path to reach Finish
const susanin = (graphLinksTree: TGraphLinksTree, startId: string, finishId: string) => {
    // track the lowest cost to reach each vertexId
    const costs = {
        [finishId]: Infinity,
        ...graphLinksTree[startId],
    };

    // track nodes that have already been processed
    const processed: string[] = [];

    // track paths
    const parents = Object.keys(graphLinksTree[startId] || {}).reduce(
        (memo, child) => {
            memo[child] = startId;
            return memo;
        },
        {
            [finishId]: ''
        }
    );

    let vertexId = getLowestCostNode(costs, processed);

    while (vertexId) {
        const currentCost = costs[vertexId];
        const children = graphLinksTree[vertexId] || {};
        Object.keys(children).forEach((n: string) => {
            if (String(n) === String(startId)) {
                // console.log('WE DON\'T GO BACK TO START', startId);
                return;
            }
            // console.log(`StartNodeName: ${startId}`);
            // console.log(`Evaluating cost to node ${n} (looking from node ${vertexId})`);
            // console.log('Last Cost:', costs[n]);

            const newCost = currentCost + children[n];
            // console.log('New Cost:', newCost);
            if (costs[n] != null && costs[n] <= newCost) {
                // console.log('A shorter path already exists');
                return;
            }

            costs[n] = newCost;
            parents[n] = vertexId;
            // console.log('Updated cost and parents');
        });
        processed.push(vertexId);
        vertexId = getLowestCostNode(costs, processed);
    }

    const optimalPath = getOptimalPath(parents, finishId);
    const cost = costs[finishId];
    return {
        cost,
        path: cost === Infinity ? [] : optimalPath,
    };
};

const WEIGHT_MAP = {
    [TVertexType.Normal]: 1,
    [TVertexType.Gravel]: 2,
};

export const isJailed = (v: TVertex) => v.type === TVertexType.Boulder;

type TOptions = {
    allowDiagonal?: boolean,
    allowPassByWormhole?: boolean,
};

export const getEdgesTree = (
    graph: TGraph,
    {
        allowDiagonal = true,
        allowPassByWormhole = false,
    }: TOptions = {},
): TGraphLinksTree => {
    const linksMap: TGraphLinksTree = {};

    const addEdgeToMap = (vFromId: TVertexId, vToId: TVertexId, weight: number | null = null) => {
        const vFrom = graph.vertices[vFromId];
        const vTo = graph.vertices[vToId];
        if (isJailed(vFrom) || isJailed(vTo)) {
            return;
        }
        if (!linksMap[vFrom.id]) {
            linksMap[vFrom.id] = {};
        }

        linksMap[vFrom.id][vTo.id] = weight === null ? WEIGHT_MAP[vFrom.type] + WEIGHT_MAP[vTo.type] : weight;
    };

    const addByCoords = (vFromId: TVertexId, x: number, y: number) => {
        let line = graph.board[y];
        const vToId: TVertexId | null = line && line[x];
        if (!vToId) {
            return;
        }
        addEdgeToMap(vFromId, vToId);
    };

    graph.board.forEach((line: TVertexId[], y: number) => {
        line.forEach((vertexId: TVertexId, x: number) => {
            const v = graph.vertices[vertexId];
            if (isJailed(v)) {
                return;
            }
            if (vertexId in graph.wormholes) {
                addEdgeToMap(vertexId, graph.wormholes[vertexId], 1);
                if (!allowPassByWormhole) {
                    return;
                }
            }
            addByCoords(vertexId, x - 1, y);
            addByCoords(vertexId, x + 1, y);
            addByCoords(vertexId, x, y - 1);
            addByCoords(vertexId, x, y + 1);
            if (allowDiagonal) {
                addByCoords(vertexId, x - 1, y - 1);
                addByCoords(vertexId, x + 1, y + 1);
                addByCoords(vertexId, x - 1, y + 1);
                addByCoords(vertexId, x + 1, y - 1);
            }
        });
    });
    return linksMap;
};

export const showMeRoute = (graph: TGraph, startId: TVertexId, finishId: TVertexId, options: TOptions = {}) => {
    const linksMap = getEdgesTree(graph, options);
    const pathResult = susanin(linksMap, startId, finishId);
    return pathResult.path;
};
