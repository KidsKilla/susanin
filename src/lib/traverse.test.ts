import { showMeRoute } from './traverse';
import { createGraphFromStringMap } from './structs';

describe('Graph traverse', () => {
    it('shold find easiest', () => {
        const graph = createGraphFromStringMap('..', []);
        const route = showMeRoute(graph, graph.board[0][0], graph.board[0][1]);
        expect(route).toMatchSnapshot();
    });
    it('shold find wormhole', () => {
        const graph = createGraphFromStringMap('.#.', [
            [[0, 0], [2, 0]]
        ]);
        const route = showMeRoute(graph, graph.board[0][0], graph.board[0][2]);
        expect(route).toMatchSnapshot();
    });
    it('shold not find', () => {
        const graph = createGraphFromStringMap('.#.', []);
        const route = showMeRoute(graph, graph.board[0][0], graph.board[0][2]);
        expect(route).toHaveLength(0);
    });
})
