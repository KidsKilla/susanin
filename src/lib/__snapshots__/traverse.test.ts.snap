// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph traverse shold find easiest 1`] = `
Array [
  "{\\"x\\":0,\\"y\\":0,\\"_\\":\\".\\"}",
  "{\\"x\\":1,\\"y\\":0,\\"_\\":\\".\\"}",
]
`;

exports[`Graph traverse shold find wormhole 1`] = `
Array [
  "{\\"x\\":0,\\"y\\":0,\\"_\\":\\".\\"}",
  "{\\"x\\":2,\\"y\\":0,\\"_\\":\\".\\"}",
]
`;
